receiver/redisreceiver/pdata.go

```diff
@@ -0,0 +1,141 @@
+	rattrs.Insert("service.name", pdata.NewAttributeValueString(serviceName))
+	ilm := pdata.NewInstrumentationLibraryMetrics()
+	rm.InstrumentationLibraryMetrics().Append(ilm)
+	ms.CopyTo(ilm.Metrics())
+	return rm
+}
+
+func buildKeyspaceTriplet(k *keyspace, t *timeBundle) pdata.MetricSlice {
+	keysPDM := buildKeyspaceKeysMetric(k, t)
+	expiresPDM := buildKeyspaceExpiresMetric(k, t)
+	ttlPDM := buildKeyspaceTTLMetric(k, t)
+
+	ms := pdata.NewMetricSlice()
+	ms.Append(keysPDM)
+	ms.Append(expiresPDM)
+	ms.Append(ttlPDM)
+
+	return ms
+}
+
+func buildKeyspaceKeysMetric(k *keyspace, t *timeBundle) pdata.Metric {
+	m := &redisMetric{
+		name:   "redis/db/keys",
+		labels: map[string]string{"db": k.db},
+		pdType: pdata.MetricDataTypeIntGauge,
+	}
+	value := int64(k.keys)
+
+	pt := pdata.NewIntDataPoint()
+	pt.SetValue(value)
+	pdm := newIntMetric(m, pt, t)
+
+	return pdm
+}
+
+func buildKeyspaceExpiresMetric(k *keyspace, t *timeBundle) pdata.Metric {
+	m := &redisMetric{
+		name:   "redis/db/expires",
+		labels: map[string]string{"db": k.db},
+		pdType: pdata.MetricDataTypeIntGauge,
+	}
+	value := int64(k.expires)
+
+	pt := pdata.NewIntDataPoint()
+	pt.SetValue(value)
+	pdm := newIntMetric(m, pt, t)
+
+	return pdm
+}
+
+func buildKeyspaceTTLMetric(k *keyspace, t *timeBundle) pdata.Metric {
+	m := &redisMetric{
+		name:   "redis/db/avg_ttl",
+		units:  "ms",
+		labels: map[string]string{"db": k.db},
+		pdType: pdata.MetricDataTypeIntGauge,
+	}
+	value := int64(k.avgTTL)
+
+	pt := pdata.NewIntDataPoint()
+	pt.SetValue(value)
+	pdm := newIntMetric(m, pt, t)
+
+	return pdm
+}
+
+func newIntMetric(m *redisMetric, pt pdata.IntDataPoint, t *timeBundle) pdata.Metric {
+	pdm := redisMetricToPDM(m)
+
+	pt.SetTimestamp(pdata.TimestampUnixNano(t.current.UnixNano()))
+	pt.LabelsMap().InitFromMap(m.labels)
+
+	var points pdata.IntDataPointSlice
+	if m.pdType == pdata.MetricDataTypeIntGauge {
+		points = pdm.IntGauge().DataPoints()
+	} else if m.pdType == pdata.MetricDataTypeIntSum {
+		pt.SetStartTime(pdata.TimestampUnixNano(t.serverStart.UnixNano()))
+		sum := pdm.IntSum()
+		sum.SetIsMonotonic(m.isMonotonic)
+		sum.SetAggregationTemporality(pdata.AggregationTemporalityCumulative)
+		points = sum.DataPoints()
+	}
+	points.Append(pt)
+
+	return pdm
+}
+
+func newDoubleMetric(m *redisMetric, pt pdata.DoubleDataPoint, t *timeBundle) pdata.Metric {
+	pdm := redisMetricToPDM(m)
+
+	pt.SetTimestamp(pdata.TimestampUnixNano(t.current.UnixNano()))
+	pt.LabelsMap().InitFromMap(m.labels)
+
+	var points pdata.DoubleDataPointSlice
+	if m.pdType == pdata.MetricDataTypeDoubleGauge {
+		points = pdm.DoubleGauge().DataPoints()
+	} else if m.pdType == pdata.MetricDataTypeDoubleSum {
+		pt.SetStartTime(pdata.TimestampUnixNano(t.serverStart.UnixNano()))
+		sum := pdm.DoubleSum()
+		sum.SetIsMonotonic(m.isMonotonic)
+		sum.SetAggregationTemporality(pdata.AggregationTemporalityCumulative)
+		points = sum.DataPoints()
+	}
+	points.Append(pt)
```

Comment:
It is better to avoid "Append" and actually create the `DataPoints()` then `Resize` it to `1` then fill that point instead of creating the point then appending. This will remove 1 unnecessary allocation.

This can be done in a followup PR.

PS: Same can be done in in a bunch of other places.

Reply:
Ah, good to know. I will do this in a followup PR.